feat: adding sequelize along with new db type 'ArbyOrder' (#85)

rsercano opened the pull request:
Hello,

this PR implements the following changes:

- Added a new db type along with db connection `ArbyOrder`
- Fixed a small issue on `.env-example `(`TEST_MODE` instead of `LIVE_CEX`)
- Changed initialization flow to add db initialization into `arby.ts`
- Created a new log context `db` and used for sequelize

Next step:

- Using `ArbyOrder` model in `mergeMap`'s `project` [here](https://github.com/ExchangeUnion/market-maker-bot/compare/feat/adding-sequelize?expand=1#diff-77a45f12e136f324d3c28fe24cdf3a531c7582103e96498a48462eaf7414a509R105): and giving it to `initCEX$`

I guess other steps (querying cex to fetch order and save) would be another PR since it would make this more than it seems. But it's up to you @erkarl if you want me to continue on this before merging, I'm okay with it too.

erkarl commented on src/db/arby-order.ts:
```diff
@@ -0,0 +1,49 @@
+    side: { type: DataTypes.TINYINT, allowNull: false },
+    quantity: { type: DataTypes.BIGINT, allowNull: false },
+    price: { type: DataTypes.BIGINT, allowNull: false },
+    time: { type: DataTypes.BIGINT, allowNull: false },
```
Time as in unix timestamp?

rsercano replied:
Yes, that's what I thought

erkarl replied:
What about https://sequelize.org/master/class/lib/data-types.js~DATE.html?

erkarl commented on src/db/arby-order.ts:
```diff
@@ -0,0 +1,49 @@
+    quantity: { type: DataTypes.BIGINT, allowNull: false },
+    price: { type: DataTypes.BIGINT, allowNull: false },
+    time: { type: DataTypes.BIGINT, allowNull: false },
+    profit: { type: DataTypes.BIGINT, allowNull: false },
```
How are you planning to populate this field?

rsercano replied:
I added what I thought here https://github.com/ExchangeUnion/market-maker-bot/pull/125#issuecomment-748606663

erkarl commented on src/db/arby-order.ts:
```diff
@@ -0,0 +1,49 @@
+  const attributes: ModelAttributes<ArbyOrderInstance> = {
+    cexOrderId: { type: DataTypes.STRING, primaryKey: true },
+    openDexOrderId: { type: DataTypes.STRING, allowNull: false },
+    side: { type: DataTypes.TINYINT, allowNull: false },
```
The order side is always both as arbitrage trade's are always asymmetrical.

rsercano replied:
It was in the issue, so I guess this's the side of CEX order as far as I understood.

erkarl reviewed (commented):
Thanks for getting started on this one.

The database structure you're suggesting to implement is a bit unclear for me. Left some comments/questions.

How are you planning to query BTC profit earned/accumulated between X and Y timestamps?

rsercano commented:
> Thanks for getting started on this one.
> 
> The database structure you're suggesting to implement is a bit unclear for me. Left some comments/questions.
> 
> How are you planning to query BTC profit earned/accumulated between X and Y timestamps?

Hi @erkarl thanks for prompt review!

In fact as far as I understood from @kilrau this profit is per trade there's no calculation between timestamps, and since we'll save this into database whenever a CEX order happens, I guess that's the case, what I had in my mind for the rest:

- CEX order executed -> 
-- Query cex order to find price/quantity/timestamp etc..
-- Find corresponding dex order to save dexOrderId.

And for profit calculation it just should be a simple minus operation, doesn't it? @kilrau @erkarl

erkarl reviewed (commented):
One DEX buy order can be linked to multiple CEX orders, and vice-versa. 

For example, if there were many DEX orders fills with very small quantity. In that case, the bot will not issue order on CEX until a configurable threshold is met.

What do you think about starting with saving all order fills (both from CEX and DEX side) to the database and timestamp it. That way we'll get the most raw data and can derive whatever metrics we want for visualization purposes and whatever timeframe.

rsercano commented:
> One DEX buy order can be linked to multiple CEX orders, and vice-versa.
> 
> For example, if there were many DEX orders fills with very small quantity. In that case, the bot will not issue order on CEX until a configurable threshold is met.
> 
> What do you think about starting with saving all order fills (both from CEX and DEX side) to the database and timestamp it. That way we'll get the most raw data and can derive whatever metrics we want for visualization purposes and whatever timeframe.

I see, so in this way we'll save whatever order is executed and how will we link them?

erkarl commented:
> and how will we link them?

There's no hard link. The query will always include start timestamp and end timestamp.

"Show me all the trades in the last 24h". We can then calculate the amounts sold/bought during the period and derive the PnL (profit and loss)

rsercano commented:
oh I see @erkarl okay if that's okay by you guys, okay for me too ofc. @kilrau @erkarl

rsercano commented:
```
export interface Order {
        id: string;
        datetime: string;
        timestamp: number;
        lastTradeTimestamp: number;
        status: 'open' | 'closed' | 'canceled';
        symbol: string;
        type: string;
        side: 'buy' | 'sell';
        price: number;
        average?: number;
        amount: number;
        filled: number;
        remaining: number;
        cost: number;
        trades: Trade[];
        fee: Fee;
        info: any;
    }
```

After the call with @erkarl and @kilrau we'll save the order as is with linked `Trades` in the database.

rsercano commented:
As an outcome to our discussion, 

- I changed ArbyOrder name to `CexOrder`
- Created two new db types `CexFee` & `CexTrade`
- `CexFee` has one to one relation with both `CexTrade` & `CexOrder`
- `CexOrder` has one to many relation with `CexTrade`

Just as in the cctx. 

Next step:

- Saving cex order as soon as it's executed.

erkarl commented on src/db/db.ts:
```diff
@@ -0,0 +1,81 @@
+    dialect: 'sqlite',
+  });
+
+  return of(createModels(sequelize)).pipe(
```
What's going on here? Did you mean to use `from` operator here instead?

rsercano replied:
Uhm, there's actually a sequence here that has to be executed sequentially, so first authenticate then model's sync promises in the given order. I changed `of` to `from` for promises but still can't see a way to shorten it.

rsercano replied:
Any advice would be appreciated

erkarl commented on src/db/db.ts:
```diff
@@ -0,0 +1,81 @@
+    foreignKey: 'orderId',
+  });
+
+  models.CexFee.belongsTo(models.CexOrder, {
```
Why is `CexFee` linked to both `CexOrder` and `CexTrade`?

rsercano replied:
Uhm, actually this's from ccxt, it's in both Trade and Order

rsercano replied:
I'm removing fee from Trade nonetheless and keeping it for Order only.

erkarl commented on src/db/cex-fee.ts:
```diff
@@ -0,0 +1,41 @@
+  cost: number;
+};
+
+type CexFeeAttributes = CexFeeCreationAttributes;
```
Why not use `CexFeeCreationAttributes` directly?

erkarl reviewed (commented):
Thanks for moving forward with this.

In general I think we can lose the `Cex` prefix for database fields `CexOrder -> Order`, `CexTrade -> Trade`, `CexFee` -> `Fee`.

I'm also a bit unsure about `Fee` in a separate table. Going to think about it for a while.

rsercano commented:
So as per your comment @erkarl I suppose while saving fee into Order, I guess it's easier to do when we don't separate Fee as you said, I'm going to change it and put fee fields into order.

rsercano commented:
Hi @erkarl what I've added `saveOrder$` observable to executeOrder of Cex, could you please check? To be honest I couldn't test this %100 since it's not being used on test mode. 

Also I added a TODO to orderRepository to use schedulers of rxjs to make save operation async, wdyt @erkarl ?

erkarl commented:
> To be honest I couldn't test this %100 since it's not being used on test mode.

You definitely should be able to test this manually.

erkarl commented on src/db/order-repository.ts:
```diff
@@ -0,0 +1,88 @@
+    const arbyOrder = this.convertToArbyOrder(order);
+    const arbyTrades = this.convertToArbyTrades(order.trades, order.id);
+    // TODO async
+    return from(this.models.Order.create(arbyOrder)).pipe(
```
Why do we need to use external `this.models` state here? Perhaps we can make due without introducing `OrderRepository` class and pass in the models as an argument?

rsercano commented:
> > To be honest I couldn't test this %100 since it's not being used on test mode.
> 
> You definitely should be able to test this manually.

Uhm, can you point me to right direction I literally never saw a log message about centralized order, I saw you are logging either way it's test or not, but I never ended up in the logs with one.

erkarl commented on src/db/db.ts:
```diff
@@ -0,0 +1,69 @@
+let sequelize: Sequelize;
+
+const initDB$ = ({
+  logger,
```
Let's change the object argument to individual arguments. This is the pattern I'd like to project to move towards for all functions so that we can provide some of the arguments by currying the function: https://ramdajs.com/docs/#curry

rsercano replied:
As we discussed on the call, we'll do that after a whole refactor.

rsercano commented:
Can you please check this once more @erkarl there was an issue because I was calling closeDB$ in cleanup but that's not right since cleanup has to happen during shutdown and has to happen once not on cleanup.

rsercano commented:
fyi. @erkarl I'll create a new branch over this one and will start HTTP implementation

erkarl commented:
> fyi. @erkarl I'll create a new branch over this one and will start HTTP implementation

@rsercano did you test this PR with real environment?

rsercano commented:
> > fyi. @erkarl I'll create a new branch over this one and will start HTTP implementation
> 
> @rsercano did you test this PR with real environment?

I tested by calling `saveOrder$ subscribe` with a dummy order, not with a real environment unfortunately.

erkarl commented on src/arby.ts:
```diff
@@ -155,15 +174,26 @@ if (!module.parent) {
         console.log(error);
       }
-      process.exit(1);
+      closeDB$().subscribe({
```
Whenever you need to subscribe within a subscribe callback you should probably be using a transformation operation instead (such as `mergeMap`).

rsercano replied:
I changed this by adding a pipe to startArby observable and then calling closeDB observable within mergeMap, but I'm really unsure if it worked, because if I put console.log statements inside to mergeMap (that I call closeDB), they're never getting called, and never printing, can you please check @erkarl 

I followed here: https://rxjs-dev.firebaseapp.com/api/operators/mergeMap

rsercano replied:
Line that I suspect is not working: https://github.com/ExchangeUnion/market-maker-bot/pull/125/files#diff-77a45f12e136f324d3c28fe24cdf3a531c7582103e96498a48462eaf7414a509R181

erkarl commented on src/db/order-repository.ts:
```diff
@@ -0,0 +1,86 @@
+
+  const arbyOrder = convertToArbyOrder(order);
+  const arbyTrades = convertToArbyTrades(order.trades, order.id);
+  // TODO async
```
TODO?

rsercano replied:
Yes actually saving into database is not a big deal, but may interrupt the responsiveness of arby, so I added a TODO there to take care later and make save operation async with schedulers of rxjs, wdyt?

erkarl commented on src/trade/cleanup.spec.ts:
```diff
@@ -59,10 +60,11 @@ describe('getCleanup$$', () => {
 
   it('removes all orders on OpenDEX and CEX', () => {
-    expect.assertions(3);
+    //expect.assertions(3);
```
Why remove?

rsercano replied:
Sorry my bad.
